controllers/spaceprovisionerconfig/spaceprovisionerconfig_controller.go

```diff
@@ -58,62 +74,180 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 	var toolchainPresent corev1.ConditionStatus
-	toolchainPresenceReason := toolchainv1alpha1.SpaceProvisionerConfigValidReason
-	var reportedError error
-	toolchainPresenceMessage := ""
-	if err := r.Client.Get(ctx, toolchainClusterKey, toolchainCluster); err != nil {
+	var err error
+	if err = r.Client.Get(ctx, toolchainClusterKey, toolchainCluster); err != nil {
 		if !errors.IsNotFound(err) {
-			// we need to requeue the reconciliation in this case because we cannot be sure whether the ToolchainCluster
-			// is really present in the cluster or not. If we did not do that and instead just reported the error in
-			// the status, we could eventually leave the SPC in an incorrect state once the error condition in the cluster,
-			// that prevents us from reading the ToolchainCluster, clears. I.e. we need the requeue to keep the promise
-			// of eventual consistency.
-
-			reportedError = fmt.Errorf("failed to get the referenced ToolchainCluster: %w", err)
-			toolchainPresenceMessage = reportedError.Error()
+			// IsNotFound is self-explanatory but let's add a little bit of context to the error
+			// in other cases
+			err = fmt.Errorf("failed to get the referenced ToolchainCluster: %w", err)
+			toolchainPresent = corev1.ConditionUnknown
+		} else {
+			toolchainPresent = corev1.ConditionFalse
 		}
-		toolchainPresenceReason = toolchainv1alpha1.SpaceProvisionerConfigToolchainClusterNotFoundReason
-		toolchainPresent = corev1.ConditionFalse
 	} else {
 		readyCond, found := condition.FindConditionByType(toolchainCluster.Status.Conditions, toolchainv1alpha1.ConditionReady)
 		if !found {
 			toolchainPresent = corev1.ConditionFalse
 		} else {
 			toolchainPresent = readyCond.Status
 		}
-		if toolchainPresent != corev1.ConditionTrue {
-			toolchainPresenceReason = toolchainv1alpha1.SpaceProvisionerConfigToolchainClusterNotReadyReason
+	}
+
+	return toolchainPresent, err
+}
+
+func (r *Reconciler) determineCapacityReadyState(spc *toolchainv1alpha1.SpaceProvisionerConfig) corev1.ConditionStatus {
+	if spc.Status.ConsumedCapacity == nil {
+		return corev1.ConditionUnknown
+	}
+
+	var freeSpaces corev1.ConditionStatus
+	max := spc.Spec.CapacityThresholds.MaxNumberOfSpaces
+	if max == 0 || max > uint(spc.Status.ConsumedCapacity.SpaceCount) {
+		freeSpaces = corev1.ConditionTrue
+	} else {
+		freeSpaces = corev1.ConditionFalse
+	}
+
+	enoughMemory := corev1.ConditionUnknown // let the state be unknown if we have no information
+
+	if spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent == 0 { // unlimited
+		enoughMemory = corev1.ConditionTrue
+	} else if len(spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole) > 0 { 
+		enoughMemory = corev1.ConditionTrue
+		for _, val := range spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole {
+			if uint(val) >= spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent {
+				enoughMemory = corev1.ConditionFalse
+				break
+			}
+		}
+	}
+
+	return And(freeSpaces, enoughMemory)
```

Comment:
this whole code looks a bit over-complicated to me. couldn't this simplified version work too?
```go
	if spc.Status.ConsumedCapacity == nil ||
		len(spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole) == 0 {
		return corev1.ConditionUnknown
	}

	max := spc.Spec.CapacityThresholds.MaxNumberOfSpaces
	if max != 0 && max <= uint(spc.Status.ConsumedCapacity.SpaceCount) {
		return corev1.ConditionFalse
	}

	if spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent != 0 {
		for _, val := range spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole {
			if uint(val) >= spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent {
				return corev1.ConditionFalse
			}
		}
	}

	return corev1.ConditionTrue
```

Reply:
I'm a divide & conquer guy :) For me, it's simpler to reason about the space count first, then the memory usage and then combine the two results together.

Your code evaluates the two of them (almost) together. Of course, it is shorter but especially the first condition is non-trivial to understand.

At least to me.

Reply:
easy to fix:
```go
if spc.Status.ConsumedCapacity == nil {
	return corev1.ConditionUnknown
}

max := spc.Spec.CapacityThresholds.MaxNumberOfSpaces
if max != 0 && max <= uint(spc.Status.ConsumedCapacity.SpaceCount) {
	return corev1.ConditionFalse
}

if len(spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole) == 0 {
	return corev1.ConditionUnknown
}

if spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent != 0 {
	for _, val := range spc.Status.ConsumedCapacity.MemoryUsagePercentPerNodeRole {
		if uint(val) >= spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent {
			return corev1.ConditionFalse
		}
	}
}

return corev1.ConditionTrue
```
if you add a comment or two, then it's like reading a story.

But anyway, let's clarify what we want to achieve: 
* if there is either no capacity at all, or it's missing the `MemoryUsagePercentPerNodeRole`, then the status should be `unknown`
* if the number of spaces is equal or more than the threshold, then the status should be `false`
* if the memory consumption of any node is higher or equal to the threshold, then the status should be `false`
* `true` in all other cases

Is that correct? If so, then the code exactly follows the expectations.

Reply:
I've updated the code. Please have another look.